python/nav/dhcpstats/common.py

```diff
@@ -0,0 +1,216 @@
+        True
+        """
+        parts = graphite_path.split(".")
+        if not len(parts) >= 9:
+            raise ValueError(
+                f"Expected graphite_path {graphite_path!r} to have at least 9 dot-separated segments"
+            )
+        ip_version = parts[2]
+        server_name = parts[3]
+        allocation_type = parts[4]
+        group_name_source = parts[5]
+        group_name = parts[6]
+        first_ip = parts[7]
+        last_ip = parts[8]
+
+        allowed_group_sources = ("special_groups", "custom_groups")
+        if group_name_source not in allowed_group_sources:
+            raise ValueError(f"group_source_name {group_name_source!r} is not in {allowed_group_sources!r}")
+
+        allowed_allocation_types = ("range", "pool", "subnet")
+        if allocation_type not in allowed_allocation_types:
+            raise ValueError(f"allocation_type {allocation_type!r} is not in {allowed_allocation_types!r}")
+
+        first_ip = cls._unescape_graphite_address(first_ip)
+        last_ip = cls._unescape_graphite_address(last_ip)
+        cls._check_ip_pair(first_ip, last_ip)
+
+        if str(first_ip.version()) != ip_version or str(last_ip.version()) != ip_version:
+            raise ValueError(f"first_ip {first_ip!r} or last_ip {last_ip!r} not of same version as expected ip_version {ip_version!r}")
+
+        return cls(
+            ip_version=first_ip.version(),
+            server_name=server_name,
+            allocation_type=allocation_type,
+            group_name_source=group_name_source,
+            group_name=group_name,
+            first_ip=first_ip,
+            last_ip=last_ip,
+        )
+
+    @classmethod
+    def from_external_info(
+            cls,
+            server_name: str,
+            allocation_type: Literal["range", "pool", "subnet"],
+            group_name: Optional[str],
+            first_ip: Union[str, IPy.IP],
+            last_ip: Union[str, IPy.IP],
+    ):
+        """
+        Instantiate me with path data sourced from an external source such as a
+        DHCP server.
+
+        if group_name is missing, group_name_source is set to "special_groups"
+         and group_name is set to "standalone" such that the returned instance
+         will be treated as belonging to a group of size one.
+        otherwise, group_name_source is set to "custom_groups".
+
+        if first_ip cannot be parsed to an IP address, raises a ValueError
+        if last_ip cannot be parsed to an IP address, raises a ValueError
+        if first_ip > last_ip, raises a ValueError
+        otherwise, returns the Path instance.
+        """
+        if group_name is None:
+            group_name_source = "special_groups"
+            group_name = "standalone"
+        else:
+            group_name_source = "custom_groups"
+            group_name = group_name
+
+        first_ip = IPy.IP(IPy.IP(first_ip)[0])
+        last_ip = IPy.IP(IPy.IP(last_ip)[-1])
```

Comment:
This should be unnecessary. Inputting prefixes into the `first_ip` and `last_ip` arguments should be considered an error, rather than trying to do some weird guesstimation and allow it...

Reply:
I'd say this is in the blocking category (these two lines just look weird)